etcd3/aio/client.py

```diff
@@ -0,0 +1,1058 @@
+        self.version = version
+        self.db_size = db_size
+        self.leader = leader
+        self.raft_index = raft_index
+        self.raft_term = raft_term
+
+
+class Alarm(object):
+    def __init__(self, alarm_type, member_id):
+        self.alarm_type = alarm_type
+        self.member_id = member_id
+
+
+class EtcdTokenCallCredentials(grpc.AuthMetadataPlugin):
+    """Metadata wrapper for raw access token credentials."""
+
+    def __init__(self, access_token):
+        self._access_token = access_token
+
+    def __call__(self, context, callback):
+        metadata = (('token', self._access_token),)
+        callback(metadata, None)
+
+
+class Etcd3Client(object):
+    def __init__(self, host='localhost', port=2379,
+                 ca_cert=None, cert_key=None, cert_cert=None, timeout=None,
+                 user=None, password=None, grpc_options=None,
+                 loop=None):
+
+        self._url = '{host}:{port}'.format(host=host, port=port)
+        self.metadata = None
+
+        cert_params = [c is not None for c in (cert_cert, cert_key)]
+        if ca_cert is not None:
+            if all(cert_params):
+                credentials = self._get_secure_creds(
+                    ca_cert,
+                    cert_key,
+                    cert_cert
+                )
+                self.uses_secure_channel = True
+                self.channel = aiogrpc.secure_channel(self._url, credentials,
+                                                      options=grpc_options,
+                                                      loop=loop)
+            elif any(cert_params):
+                # some of the cert parameters are set
+                raise ValueError(
+                    'to use a secure channel ca_cert is required by itself, '
+                    'or cert_cert and cert_key must both be specified.')
+            else:
+                credentials = self._get_secure_creds(ca_cert, None, None)
+                self.uses_secure_channel = True
+                self.channel = aiogrpc.secure_channel(self._url, credentials,
+                                                      options=grpc_options,
+                                                      loop=loop)
+        else:
+            self.uses_secure_channel = False
+            self.channel = aiogrpc.insecure_channel(self._url,
+                                                    options=grpc_options,
+                                                    loop=loop)
+
+        self.timeout = timeout
+        self.call_credentials = None
+
+        cred_params = [c is not None for c in (user, password)]
+
+        if all(cred_params):
+            self.auth_stub = etcdrpc.AuthStub(self.channel)
+            auth_request = etcdrpc.AuthenticateRequest(
+                name=user,
+                password=password
+            )
+
+            resp = self.auth_stub.Authenticate(auth_request, self.timeout)
```

Comment:
is this a blocking call?

Reply:
there's a whole bunch of blocking IO calls in this constructor - perhaps it would be better to move those to a lazy call in `__aenter__`

```python
async def channel():
    if self._channel is not None:
        return self._channel
    ...
    credentials = await self.__get_secure_credentials(...)
    ...
    self._channel = ...
    return self._channel

async def __aenter__(self):
    await self.channel()
    return self
```

and then call `await self.channel()` where you used to call `self.channel`